crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,160 @@
+}
+
+impl Default for Definitions {
+    fn default() -> Self {
+        Self {
+            definitions: 0..0,
+            may_be_unbound: true,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct FlowSnapshot {
+    definitions_by_symbol: IndexVec<ScopedSymbolId, Definitions>,
+}
+
+pub(super) struct UseDefMapBuilder<'db> {
+    all_definitions: Vec<Definition<'db>>,
+
+    definitions_by_use: IndexVec<ScopedUseId, Definitions>,
+
+    // builder state: currently visible definitions for each symbol
+    definitions_by_symbol: IndexVec<ScopedSymbolId, Definitions>,
+}
+
+impl<'db> UseDefMapBuilder<'db> {
+    pub(super) fn new() -> Self {
+        Self {
+            all_definitions: Vec::new(),
+            definitions_by_use: IndexVec::new(),
+            definitions_by_symbol: IndexVec::new(),
+        }
+    }
+
+    pub(super) fn add_symbol(&mut self, symbol: ScopedSymbolId) {
+        let new_symbol = self.definitions_by_symbol.push(Definitions::default());
+        debug_assert_eq!(symbol, new_symbol);
+    }
+
+    pub(super) fn record_def(&mut self, symbol: ScopedSymbolId, definition: Definition<'db>) {
+        let def_idx = self.all_definitions.len();
+        self.all_definitions.push(definition);
+        self.definitions_by_symbol[symbol] = Definitions {
+            #[allow(clippy::range_plus_one)]
+            definitions: def_idx..(def_idx + 1),
+            may_be_unbound: false,
+        };
+    }
+
+    pub(super) fn record_use(&mut self, symbol: ScopedSymbolId, use_id: ScopedUseId) {
+        let new_use = self
+            .definitions_by_use
+            .push(self.definitions_by_symbol[symbol].clone());
+        debug_assert_eq!(use_id, new_use);
+    }
+
+    pub(super) fn snapshot(&self) -> FlowSnapshot {
+        FlowSnapshot {
+            definitions_by_symbol: self.definitions_by_symbol.clone(),
+        }
+    }
+
+    pub(super) fn set(&mut self, state: &FlowSnapshot) {
+        let num_symbols = self.definitions_by_symbol.len();
+        self.definitions_by_symbol = state.definitions_by_symbol.clone();
+        self.definitions_by_symbol
+            .resize(num_symbols, Definitions::default());
+    }
+
+    pub(super) fn merge(&mut self, state: &FlowSnapshot) {
+        for (symbol_id, to_merge) in state.definitions_by_symbol.iter_enumerated() {
+            let current = self.definitions_by_symbol.get_mut(symbol_id).unwrap();
+            // if the symbol can be unbound in either predecessor, it can be unbound
+            current.may_be_unbound |= to_merge.may_be_unbound;
+            // merge the definition ranges
+            if current.definitions == to_merge.definitions {
+                // ranges already identical, nothing to do!
+            } else if current.definitions.end == to_merge.definitions.start {
+                // ranges adjacent (current first), just merge them
+                current.definitions = (current.definitions.start)..(to_merge.definitions.end);
```

Comment:
The parentheses here and in other branches are unnecessary
```suggestion
                current.definitions = current.definitions.start..to_merge.definitions.end;
```

Reply:
I personally find it harder to visually parse without the parens, but if this isn't common style, I don't mind removing the parens.